feat: ocr

nyakang opened the pull request:
## Description

feat: add OCR functionality and related configurations

- Added new OCR-related translations in English and Simplified Chinese
- Integrated PaddleOCR for text recognition in images
- Enhanced search capabilities by adding an “OCR” search type
- Implemented backend support for OCR processing, including database schema updates and service integration
- Updated frontend components to support OCR settings and search filters, including mobile OCR search-filter functionality


## How Has This Been Tested?

- [x] Tested on a **fresh full deployment**
- [x] Tested on an **incremental / rolling upgrade**


<details><summary><h2>Screenshots (if appropriate)</h2></summary>

![image](https://github.com/user-attachments/assets/f6ec1742-c1a5-450f-884f-2c2a6ed0fc13)

https://github.com/user-attachments/assets/d61cd4ed-57fa-4518-abee-d0cb577c31a4

https://github.com/user-attachments/assets/4ed37925-a068-42be-abbf-cac0be2b97f7

</details>


## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if applicable
- [x] I have no unrelated changes in the PR.
- [x] I have confirmed that any new dependencies are strictly necessary.
- [x] I have written tests for new code (if applicable)
- [x] I have followed naming conventions/patterns in the surrounding code
- [x] All code in `src/services/` uses repositories implementations for database calls, filesystem operations, etc.
- [x] All code in `src/repositories/` is pretty basic/simple and does not have any immich specific logic (that belongs in `src/services/`)

nyakang committed:
feat: add OCR functionality and related configurations

mertalev commented on machine-learning/immich_ml/models/constants.py:
```diff
@@ -75,6 +75,10 @@
 
 
+_PADDLE_MODELS = {
+    "paddle",
```
PaddleOCR is a hub of many models, not just one. The name should be the actual model being used (I think one of the ppocr variants is default).

mertalev commented on machine-learning/immich_ml/models/ocr/paddle.py:
```diff
@@ -0,0 +1,47 @@
+        self.loaded = True
+
+    def _load(self) -> None:
+        try:
```
This try/catch is unnecessary

mertalev commented on machine-learning/immich_ml/models/ocr/paddle.py:
```diff
@@ -0,0 +1,47 @@
+    def _load(self) -> None:
+        try:
+            self.model = PaddleOCR(
+                use_doc_orientation_classify=False,
+                use_doc_unwarping=False,
+                use_textline_orientation=False
```
What do these settings do? Also, does this set a resolution for the input or does it run at native resolution?

mertalev commented on machine-learning/immich_ml/models/ocr/paddle.py:
```diff
@@ -0,0 +1,47 @@
+            (text, score)
+            for result in results
+            for text, score in zip(result['rec_texts'], result['rec_scores'])
+            if score > self.min_score
```
Assuming the output is an ndarray, use its filtering syntax instead (e.g. `results["rec_texts"][results["rec_scores"] > self.min_score]`)

mertalev commented on machine-learning/immich_ml/models/ocr/paddle.py:
```diff
@@ -0,0 +1,47 @@
+            for text, score in zip(result['rec_texts'], result['rec_scores'])
+            if score > self.min_score
+        ]
+        if not valid_texts_and_scores:
+            return OCROutput(text="", confidence=0.0)
+        texts, scores = zip(*valid_texts_and_scores)
+        return OCROutput(
+            text="".join(texts),
+            confidence=sum(scores) / len(scores)
+        )
```
Don't concatenate them - the bounding box positions are valuable information. The table should store them in addition to the text itself as well.

mertalev commented on machine-learning/pyproject.toml:
```diff
@@ -47,11 +49,15 @@ lint = [
-openvino = ["onnxruntime-openvino>=1.17.1,<1.19.0"]
-armnn = ["onnxruntime>=1.15.0,<2"]
-rknn = ["onnxruntime>=1.15.0,<2", "rknn-toolkit-lite2>=2.3.0,<3"]
+cpu = ["onnxruntime>=1.15.0,<2", "paddlepaddle==3.0.0rc1"]
+cuda = ["onnxruntime-gpu>=1.17.0,<2", "paddlepaddle-gpu==3.0.0rc1"]
+openvino = ["onnxruntime-openvino>=1.17.1,<1.19.0", "paddlepaddle==3.0.0rc1"]
+armnn = ["onnxruntime>=1.15.0,<2", "paddlepaddle==3.0.0rc1"]
+rknn = [
+    "onnxruntime>=1.15.0,<2",
+    "rknn-toolkit-lite2>=2.3.0,<3",
+    "paddlepaddle==3.0.0rc1",
+]
 rocm = []
```
It's kind of unfortunate that only cuda is accelerated.

mertalev commented on server/src/controllers/search.controller.ts:
```diff
@@ -43,6 +44,13 @@ export class SearchController {
     return this.service.searchSmart(auth, dto);
   }
 
+  @Post('ocr')
+  @HttpCode(HttpStatus.OK)
+  @Authenticated()
+  searchOcr(@Auth() auth: AuthDto, @Body() dto: OcrSearchDto): Promise<SearchResponseDto> {
+    return this.service.searchOcr(auth, dto);
+  }
```
Does this really need to be its own endpoint? It'd be more powerful as a field in the other search endpoints.

mertalev commented on server/src/entities/ocr.entity.ts:
```diff
@@ -0,0 +1,5 @@
+export class OcrEntity {
+  id!: string;
+  assetId!: string;
+  text!: string;
```
The server no longer has entities. It should be a Table under `schema/tables`, similar to the others.

mertalev commented on server/src/migrations/1743429240851-UpsertOcrAssetJobStatus.ts:
```diff
@@ -0,0 +1,19 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class UpsertOcrAssetJobStatus1743429240851 implements MigrationInterface {
```
We no longer use TypeORM for migrations. Update the table in `schema/tables` and generate a Kysely migration instead (I believe the command is `npm run migrations:generate UpsertOcrAssetJobStatus`).

mertalev commented on server/src/repositories/asset-job.repository.ts:
```diff
@@ -342,6 +342,23 @@ export class AssetJobRepository {
+    .select(['assets.id'])
+    .$if(!force, (qb) =>
+      qb
+        .leftJoin('asset_job_status', 'asset_job_status.assetId', 'assets.id')
+        .where((eb) =>
+          eb.or([eb('asset_job_status.ocrAt', 'is', null), eb('asset_job_status.assetId', 'is', null)]),
+        )
```
Make this an inner join and remove the second condition in the where clause.

mertalev commented on server/src/repositories/ocr.repository.ts:
```diff
@@ -0,0 +1,74 @@
+  constructor(@InjectKysely() private db: Kysely<DB>) {}
+
+  @GenerateSql({ params: [DummyValue.UUID] })
+  getOcrById(id: string): Promise<OcrEntity | null> {
```
Omit the response types and don't cast the responses in these methods; the types are more precise that way. Also, these methods don't need to mention `Ocr` - that's already implied by this being the ocr repository.

mertalev commented on server/src/repositories/ocr.repository.ts:
```diff
@@ -0,0 +1,74 @@
+    return this.db
+      .selectFrom('asset_ocr')
+      .selectAll('asset_ocr')
+      .$if(!!options.assetId, (qb) => qb.where('asset_ocr.assetId', '=', options.assetId!))
```
I don't think this really makes sense to have in a stream method.

mertalev commented on server/src/repositories/ocr.repository.ts:
```diff
@@ -0,0 +1,74 @@
+    return this.db
+      .selectFrom('asset_ocr')
+      .selectAll('asset_ocr')
+      .where('text', 'is', null)
```
I don't think this should happen to begin with. If there's no text, don't add anything to this table. The text column should have a non-null constraint in the DB as well.

mertalev commented on server/src/repositories/ocr.repository.ts:
```diff
@@ -0,0 +1,74 @@
+      .execute() as Promise<OcrEntity[]>;
+  }
+
+  async delete(ocr: OcrEntity[]): Promise<void> {
```
Change the signature to accept an array of ids instead.

mertalev commented on server/src/repositories/ocr.repository.ts:
```diff
@@ -0,0 +1,74 @@
+
+
+  @GenerateSql()
+  async getLatestOcrDate(): Promise<string | undefined> {
```
This only exists in the person repository because there's a nightly job that uses it. You should remove it if you don't need it.

mertalev commented on server/src/repositories/search.repository.ts:
```diff
@@ -250,6 +257,35 @@ export class SearchRepository {
+      .where('asset_ocr.text', 'ilike', `%${options.ocr}%`)
+      .limit(pagination.size + 1)
+      .offset((pagination.page - 1) * pagination.size)
+      .execute() as any;
```
Remove the cast.

nyakang committed:
chore: update labeler configuration for machine learning files

nyakang committed:
feat(i18n): enhance OCR model descriptions and add orientation classification and unwarping features

nyakang committed:
chore: update Dockerfile to include ccache for improved build performance

nyakang committed:
feat(ocr): enhance OCR model configuration with orientation classification and unwarping options, update PaddleOCR integration, and improve response structure

nyakang committed:
refactor(ocr): remove OCR_CLEANUP job from enum and type definitions

nyakang committed:
refactor(ocr): remove obsolete OCR entity and migration files, and update asset job status and schema to accommodate new OCR table structure

nyakang committed:
refactor(ocr): update OCR schema and response structure to use individual coordinates instead of bounding box, and adjust related service and repository files

flipped-1121 commented on server/src/schema/tables/asset-ocr.table.ts:
```diff
@@ -0,0 +1,30 @@
+  boundingBoxX2!: number;
+
+  @Column({ default: 0, type: 'integer' })
+  boundingBoxY2!: number;
```
Actually, I don't know what concrete purpose these coordinates serve.

mertalev commented on server/src/repositories/search.repository.ts:
```diff
@@ -250,6 +257,30 @@ export class SearchRepository {
     });
   }
 
+  @GenerateSql({
+    params: [
+      { page: 1, size: 100 },
+      {
+        userIds: [DummyValue.UUID],
+        ocr: DummyValue.STRING,
+      },
+    ],
+  })
+  async searchOcr(pagination: SearchPaginationOptions, options: OcrSearchOptions) {
+    if (!isValidInteger(pagination.size, { min: 1, max: 1000 })) {
+      throw new Error(`Invalid value for 'size': ${pagination.size}`);
+    }
+
+    const items = await searchAssetBuilder(this.db, options)
+      .innerJoin('asset_ocr', 'assets.id', 'asset_ocr.assetId')
+      .where('asset_ocr.text', 'ilike', `%${options.ocr}%`)
+      .limit(pagination.size + 1)
+      .offset((pagination.page - 1) * pagination.size)
+      .execute();
+
+    return paginationHelper(items, pagination.size);
+  }
```
Instead of having `searchOcr`, the ocr filtering should be handled in `searchAssetBuilder` so it can be used in all search methods (`searchSmart`, `searchMetadata`, `searchRandom`)

nyakang committed:
feat: enhance OCR configuration and functionality

- Updated OCR settings to include minimum detection box score, minimum detection score, and minimum recognition score.
- Refactored PaddleOCRecognizer to utilize new scoring parameters.
- Introduced new database tables for asset OCR data and search functionality.
- Modified related services and repositories to support the new OCR features.
- Updated translations for improved clarity in settings UI.

mertalev committed:
sql changes

mertalev committed:
use rapidocr

mertalev committed:
change dto

mertalev committed:
update web

mertalev committed:
update lock

mertalev committed:
update api

mertalev committed:
store positions as normalized floats

mertalev committed:
match column order in db

mertalev committed:
update admin ui settings descriptions

fix max resolution key

set min threshold to 0.1

fix bind

mertalev committed:
apply config correctly, adjust defaults

mertalev committed:
unnecessary model type

mertalev committed:
unnecessary sources

nyakang committed:
fix(ocr): switch RapidOCR lang type from LangDet to LangRec

nyakang committed:
fix(ocr): expose lang_type (LangRec.CH) and font_path on OcrOptions for RapidOCR

nyakang committed:
fix(ocr): make OCR text search case- and accent-insensitive using ILIKE + unaccent

nyakang committed:
fix(ocr): add OCR search fields

nyakang committed:
Merge branch 'main' into ocr

nyakang committed:
Merge branch 'main' into ocr

# Conflicts:
#	machine-learning/uv.lock
#	mobile/openapi/README.md

nyakang committed:
Merge branch 'main' into ocr

# Conflicts:
#	mobile/openapi/README.md

nyakang committed:
Merge branch 'main' into ocr

nyakang committed:
Merge branch 'main' into ocr

# Conflicts:
#	mobile/openapi/lib/api.dart
#	mobile/openapi/lib/api_client.dart
#	mobile/openapi/lib/model/system_config_machine_learning_dto.dart
#	open-api/immich-openapi-specs.json
#	open-api/typescript-sdk/src/fetch-client.ts
#	server/src/dtos/system-config.dto.ts

nyakang committed:
Merge branch 'main' into ocr

# Conflicts:
#	machine-learning/uv.lock

nyakang committed:
fix: Add OCR database migration and update ML prediction logic.

nyakang committed:
Merge branch 'main' into ocr

nyakang committed:
Merge remote-tracking branch 'origin/ocr' into ocr

mertalev committed:
trigrams are already case insensitive

mertalev committed:
add tests

mertalev committed:
format

mertalev committed:
update migrations

mertalev committed:
wrong uuid function

mertalev committed:
linting

mertalev committed:
maybe fix medium tests

mertalev committed:
formatting

mertalev committed:
fix weblate check

mertalev committed:
openapi

mertalev committed:
sql

mertalev committed:
minor fixes

mertalev committed:
maybe fix medium tests part 2

mertalev committed:
passing medium tests

mertalev committed:
format web

mertalev committed:
readd sql

mertalev committed:
format dart

mertalev committed:
disabled in e2e

nyakang committed:
Merge branch 'main' into ocr

# Conflicts:
#	i18n/en.json
#	mobile/openapi/README.md
#	mobile/openapi/lib/api.dart
#	mobile/openapi/lib/api_client.dart

dodysw3 commented:
Hi on this fork, when using Nvidia CUDA on a 12GB vram, I noticed after a while, the gpu run out of memory:

immich_machine_learning  |                              ONNXRuntimeError: Traceback (most recent call
immich_machine_learning  |                              last):
immich_machine_learning  |                                File
immich_machine_learning  |                              "/opt/venv/lib/python3.11/site-packages/rapidocr/in
immich_machine_learning  |                              ference_engine/onnxruntime/main.py", line 90, in
immich_machine_learning  |                              __call__
immich_machine_learning  |                                  return
immich_machine_learning  |                              self.session.run(self.get_output_names(),
immich_machine_learning  |                              input_dict)[0]
immich_machine_learning  |                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
immich_machine_learning  |                              ^^^^^^^^^^^^^
immich_machine_learning  |                                File
immich_machine_learning  |                              "/opt/venv/lib/python3.11/site-packages/onnxruntime
immich_machine_learning  |                              /capi/onnxruntime_inference_collection.py", line
immich_machine_learning  |                              220, in run
immich_machine_learning  |                                  return self._sess.run(output_names, input_feed,
immich_machine_learning  |                              run_options)
immich_machine_learning  |                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
immich_machine_learning  |                              ^^^^^^^^^^^^^
immich_machine_learning  |                              onnxruntime.capi.onnxruntime_pybind11_state.Runtime
immich_machine_learning  |                              Exception: [ONNXRuntimeError] : 6 :
immich_machine_learning  |                              RUNTIME_EXCEPTION : Non-zero status code returned
immich_machine_learning  |                              while running Resize node. Name:'Resize.6' Status
immich_machine_learning  |                              Message:
immich_machine_learning  |                              /onnxruntime_src/onnxruntime/core/framework/bfc_are
immich_machine_learning  |                              na.cc:376 void*
immich_machine_learning  |                              onnxruntime::BFCArena::AllocateRawInternal(size_t,
immich_machine_learning  |                              bool, onnxruntime::Stream*, bool,
immich_machine_learning  |                              onnxruntime::WaitNotificationFn) Failed to allocate
immich_machine_learning  |                              memory for requested buffer of size 802160640

I wonder if there is a known memory leak issues or too aggressive caching on RapidOCR side?

alextran1502 committed:
merge main

alextran1502 committed:
chore: translation ordering

jrasm91 reviewed (approved):
A quick look at everything and it looks great. Thanks for also writing medium tests!

alextran1502 committed:
Merge branch 'main' into ocr

dotlambda commented on machine-learning/pyproject.toml:
```diff
@@ -22,6 +22,8 @@ dependencies = [
     "rich>=13.4.2",
     "tokenizers>=0.15.0,<1.0",
     "uvicorn[standard]>=0.22.0,<1.0",
+    "setuptools>=78.1.0",
```
I don't think this is used anywhere.
```suggestion
```
